File: traversal/in-order/recursion.ts
import { BinaryTreeNode } from 'data-structures/src/binary_tree'

export function inOrderTraversal<V>(node: BinaryTreeNode<V> | null): V[] {
  const answer: V[] = []
  traversal(node, answer)
  return answer
}

function traversal<V>(node: BinaryTreeNode<V> | null, answer: V[]) {
  if (node === null) return

  traversal(node.left, answer)
  answer.push(node.value)
  traversal(node.right, answer)
}

File: traversal/in-order/recursion.go
package traversal

import structures "algorithms/data-structures/binary-tree"

// RecursiveInOrder for binary tree in-order traversal
func RecursiveInOrder(root *structures.BinaryTreeNode) []int {
	answer := []int{}
	traversal(root, &answer)
	return answer
}

func traversal(root *structures.BinaryTreeNode, answer *[]int) {
	if root == nil {
		return
	}
	traversal(root.Left, answer)
	*answer = append(*answer, root.Val.(int))
	traversal(root.Right, answer)
}
